lib/charms/traefik_k8s/v0/ingress_per_unit.py

```diff
@@ -419,10 +437,60 @@ def is_failed(self, relation: Relation = None) -> bool:
 
         return False
 
-    def get_request(self, relation: Relation):
-        """Get the IngressRequest for the given Relation."""
-        provider_app_data, requirer_unit_data = self._fetch_relation_data(relation)
-        return IngressRequest(self, relation, provider_app_data, requirer_unit_data)
+    def is_unit_ready(self, relation: Relation, unit: Unit) -> bool:
+        """Report whether the given unit has shared data in its unit data bag."""
+        # sanity check: this should not occur in production, but it may happen
+        # during testing: cfr https://github.com/canonical/traefik-k8s-operator/issues/39
+        assert (
+            unit in relation.units
+        ), "attempting to get ready state for unit that does not belong to relation"
+        if relation.data.get(unit, {}).get("data"):
+            # TODO consider doing schema-based validation here
+            return True
+        return False
+
+    def get_data(self, relation: Relation, unit: Unit, validate: bool = False) -> "RequirerData":
+        """Fetch the data shared by the specified unit on the relation (Requirer side)."""
+        data = _deserialize_data(relation.data[unit]["data"])
+        if validate:
+            _validate_data(data, INGRESS_REQUIRES_UNIT_SCHEMA)
+        return data
+
+    def publish_url(self, relation: Relation, unit_name: str, url: str):
+        """Place the ingress url in the application data bag for the units on the requires side.
+
+        Assumes that this unit is leader.
+        """
+        raw_data = relation.data[self.app].get("data", None)
+        data = _deserialize_data(raw_data) if raw_data else {"ingress": {}}
+
+        # TODO: is this necessary?
```

Comment:
This feels like a thing to iron out sooner rather than later...

Reply:
We are validating the data in the databag before adding to it; it makes sense, so I chose to keep it.
However, I thought, why should we only validate the data before, but not after? So I added an 'output' validation step, so  after touching the databag we validate again. Better to catch it here and now, before we write the data, rather than the next time we have to read from it. Felt more pythonic and debuggable.